### PR TITLE
Extracted inverse_turning_radiuses().

### PR DESCRIPTION
**Pull request by lewie-donckers**
Extracted pure function `inverse_turning_radiuses()` from `Controller::setPlan()` and added some unittests.

### CODE REVIEW

File: src/calculations.cpp [hunk 1]
```diff
@@ -40,4 +40,25 @@ std::vector<double> distances_to_goal(const std::vector<tf2::Transform> & deltas
   return result;
 }
 
+std::vector<double> inverse_turning_radiuses(const std::vector<tf2::Transform> & deltas)
+{
+  auto result = std::vector<double>{};
+
+  result.reserve(deltas.size() + 1);
```
Comment (Timple) [inline src/calculations.cpp:47]: Hah, didn't know this one yet.

File: src/calculations.cpp [hunk 2]
```diff
@@ -40,4 +40,25 @@ std::vector<double> distances_to_goal(const std::vector<tf2::Transform> & deltas
...
+  std::transform(
+    deltas.cbegin(), deltas.cend(), std::back_inserter(result), [](const tf2::Transform & d) {
```
Comment (Timple) [inline src/calculations.cpp:49]: What's the benefit of this construction versus:

```c++
for (const auto delta : deltas) {
  radius = magic;
  result.push_back(radius)
}
return result;

```

File: test/unittests/test_calculations.cpp [hunk 3]
```diff
@@ -84,4 +85,51 @@ TEST(PathTrackingPidCalculations, DistancesToGoal_IgnoreZ)
...
+TEST(PathTrackingPidCalculations, InverseTurningRadiuses_Empty)
+{
+  const auto deltas = std::vector<tf2::Transform>{};
+  const auto ref = std::vector<double>{0.0};
+  const auto result = inverse_turning_radiuses(deltas);
+
+  EXPECT_EQ(ref, result);
+}
+
+TEST(PathTrackingPidCalculations, InverseTurningRadiuses_SingleTooSmall)
```
Comment (Timple) [inline test/unittests/test_calculations.cpp:97]: Thanks for adding tests!

File: src/controller.cpp [hunk 4]
```diff
@@ -228,23 +228,7 @@ void Controller::setPlan(
   const auto deltas = deltas_of_plan(global_plan_tf_);
```
Comment (Timple) [inline src/controller.cpp:228]: I think it makes more sense to combine these two functions.

` turning_radius_inv_vector_ = inverse_turning_radiuses(global_plan_tf_)`;

This reads better. That you need delta's is an implementation detail not relevent for this part of the code.

Comment (lewie-donckers) [reply src/controller.cpp:228]: I could do that but we also need the deltas for the distances to the goal (` distance_to_goal_vector_ = distances_to_goal(deltas);`). See PR #98 . If i merge the deltas into those functions, we'd have to do that work twice.

Comment (Timple) [reply src/controller.cpp:220]: Well, you can call that function as first step in this function right?

No duplicate code in that case.

Comment (lewie-donckers) [reply src/controller.cpp:220]: No duplicate code. But now `deltas_of_plan()` is called once in `Controller::setPlan()`:

```cpp
  const auto deltas = deltas_of_plan(global_plan_tf_);

  distance_to_goal_vector_ = distances_to_goal(deltas);
  turning_radius_inv_vector_ = inverse_turning_radiuses(deltas);
```

If we inline `deltas_of_plan()` in `distances_to_goal()` and `inverse_turning_radiuses()`, that work would get done twice. I'm not saying we shouldn't, it's just something to be aware of.

Comment (Timple) [reply src/controller.cpp:220]: Ah right, let's keep it like this!

Sorry, I was only looking at the delta :)

**Review by Timple — APPROVED** [review]

**Review by Rayman — APPROVED** [review]

**Pull request merged**